tests/AMFDeserializeTests.cpp

```diff
@@ -0,0 +1,286 @@
+#include <chrono>
+#include <fstream>
+#include <iostream>
+
+#include "AMFDeserialize.h"
+#include "AMFFormat.h"
+#include "CommonCxxTests.h"
+
+AMFArrayValue* ReadFromBitStream(RakNet::BitStream* bitStream) {
+	AMFDeserialize reader;
+
+	return reader.Read(bitStream, true);
+}
+
+/**
+ * This test checks that if we recieve an unimplemented AMFValueType 
+ * we correctly throw an error and can actch it.
+ */
+void TestUnimplementedAMFValues() {
+	std::vector<AMFValueType> unimplementedValues = {
+		AMFValueType::AMFInteger,
+		AMFValueType::AMFXMLDoc,
+		AMFValueType::AMFDate,
+		AMFValueType::AMFObject,
+		AMFValueType::AMFXML,
+		AMFValueType::AMFByteArray,
+		AMFValueType::AMFVectorInt,
+		AMFValueType::AMFVectorUInt,
+		AMFValueType::AMFVectorDouble,
+		AMFValueType::AMFVectorObject,
+		AMFValueType::AMFDictionary
+		};
+	// Run unimplemented tests to check that errors are thrown if
+	// unimplemented AMF values are attempted to be parsed.
+	std::ifstream fileStream;
+	fileStream.open("AMFBitStreamUnimplementedTest.bin", std::ios::binary);
+
+	// Read a test BitStream from a file
+	std::vector<char> baseBitStream;
+	char byte = 0;
+	while (fileStream.get(byte)) {
+		baseBitStream.push_back(byte);
+	}
```

Comment:
Ill investigate some better ways to read files. Seemed good for getting the base tests across for now though.

Reply:
@Xiphoseer Isn't this probably the best way for getting it into a vector of uint8s though?

Reply:
no because it's just semantically the same (which, granted, a compiler might recognize and optimize out), but what you really want is

1. Choose a power of 2 for chunks to read at once
2. Allocate a vector with the length of the input (which you can get for a file by seeking to end and getting the offset), possibly rounded up to a multiple of your chunk size
3. Ask the file stream to read one chunk of bytes at once in a loop, which will just produce less for the last iteration.

The underlying OS is usually really good at putting a bunch of sequential bytes in a buffer with one system call, no need trigger those with increments of one by reading one byte at a time from the file stream.

Reply:
Additionally, if you go from buffer to buffer, C functions like `memcpy` already take care of optimizing small, large and (un)aligned copies efficiently.